File: prototyp/webcam/overlay.py
"""Proof screen / overlay (concept document, section 8) - an OpenCV window on monitor 2.

Shows the camera image OR the mask ('m' key), the grid, active cells, the
zone frames, the currently triggered zone highlighted and the currently
pressed key in large print. Returns the pressed control key per frame.
"""

import time

import cv2
import numpy as np

import config

WIN = "Webcam-Steuerung (Beweis-Screen)"

_KEY_LABEL = {
    config.KEYS["left"]:  "<- A",
    config.KEYS["right"]: "D ->",
    config.KEYS["up"]:    "[ SPRUNG ]",
    config.KEYS["down"]:  "v S",
}


class Overlay:
    def __init__(self):
        self.show_mask = False
        self._last_t = None
        self._fps = 0.0
        cv2.namedWindow(WIN, cv2.WINDOW_NORMAL)
        cv2.moveWindow(WIN, config.MONITOR2_X_OFFSET, 0)

    def _base_image(self, s):
        if self.show_mask and s.grid is not None:
            mask = (s.grid.astype(np.uint8) * 255)
            img = cv2.resize(mask, (config.FRAME_WIDTH, config.FRAME_HEIGHT),
                             interpolation=cv2.INTER_NEAREST)
            return cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
        if s.frame is not None:
            return s.frame.copy()
        return np.zeros((config.FRAME_HEIGHT, config.FRAME_WIDTH, 3), np.uint8)

    def show(self, s, detector, current_key, triggered_zone, send_active=False) -> str:
        img = self._base_image(s)
        h, w = img.shape[:2]

        # Draw active cells as a semi-transparent layer.
        if s.grid is not None:
            overlay_layer = img.copy()
            rows, cols = s.grid.shape
            cw, ch = w / cols, h / rows
            ys, xs = np.where(s.grid)
            for y, x in zip(ys, xs):
                p0 = (int(x * cw), int(y * ch))
                p1 = (int((x + 1) * cw), int((y + 1) * ch))
                cv2.rectangle(overlay_layer, p0, p1, (0, 200, 255), -1)
            cv2.addWeighted(overlay_layer, 0.35, img, 0.65, 0, img)

        for c in range(1, config.GRID_COLS):
            x = int(c * w / config.GRID_COLS)
            cv2.line(img, (x, 0), (x, h), (60, 60, 60), 1)
        for r in range(1, config.GRID_ROWS):
            y = int(r * h / config.GRID_ROWS)
            cv2.line(img, (0, y), (w, y), (60, 60, 60), 1)

        for name, (x0, y0, x1, y1) in config.ZONES.items():
            p0 = (int(x0 * w), int(y0 * h))
            p1 = (int(x1 * w), int(y1 * h))
            triggered = (name == triggered_zone)
            color = (0, 0, 255) if triggered else (0, 255, 0)
            thick = 4 if triggered else 1
            cv2.rectangle(img, p0, p1, color, thick)
            cv2.putText(img, name, (p0[0] + 4, p0[1] + 18),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, color, 1, cv2.LINE_AA)

        if current_key:
            label = _KEY_LABEL.get(current_key, current_key.upper())
            cv2.putText(img, label, (int(w * 0.30), 50),
                        cv2.FONT_HERSHEY_SIMPLEX, 1.4, (0, 0, 255), 3, cv2.LINE_AA)

        if send_active:
            txt, col = "SEND: AN", (0, 0, 255)
        else:
            txt, col = "SEND: AUS", (180, 180, 180)
        (tw, _), _ = cv2.getTextSize(txt, cv2.FONT_HERSHEY_SIMPLEX, 0.9, 2)
        cv2.putText(img, txt, (w - tw - 12, 36),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.9, col, 2, cv2.LINE_AA)

        now = time.monotonic()
        if self._last_t is not None:
            dt = now - self._last_t
            if dt > 0:
                self._fps = 0.9 * self._fps + 0.1 * (1.0 / dt)
        self._last_t = now

        status = (f"FPS {self._fps:4.1f} | Grid {config.GRID_COLS}x{config.GRID_ROWS} "
                  f"| enter {detector.enter_ratio:.2f} exit {detector.exit_ratio:.2f} "
                  f"| {'KALIBRIERT' if detector.calibrated else 'kalibriere...'} "
                  f"| m=Maske c=Kalib k=Send q=Ende")
        cv2.putText(img, status, (6, h - 8),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.45, (255, 255, 255), 1, cv2.LINE_AA)

        cv2.imshow(WIN, img)
        key = cv2.waitKey(1) & 0xFF
        if key in (ord("q"),):
            return "q"
        if key == 27:
            return "esc"
        if key == ord("m"):
            self.show_mask = not self.show_mask
        if key == ord("c"):
            return "c"
        if key == ord("k"):
            return "toggle_send"
        return ""

    def close(self):
        cv2.destroyAllWindows()

File: webcam-rust/src/overlay.rs
//! Beweis-Screen / Overlay auf Monitor 2 (Port von overlay.py).
//!
//! Gezeichnet wird mit OpenCV-imgproc auf den Frame; Fenster + Tasten laufen
//! ueber minifb (reines Rust, kein Qt/highgui). Tasten: 'm' Bild/Maske,
//! 'c' neu kalibrieren, 'k' Send an/aus, 'q'/ESC Ende.

use std::time::Instant;

use minifb::{Key as MKey, KeyRepeat, Window, WindowOptions};
use opencv::core::{self, Mat, Point, Scalar, Size};
use opencv::prelude::*;
use opencv::imgproc;

use crate::config;
use crate::keysender::Key;

const FONT: i32 = imgproc::FONT_HERSHEY_SIMPLEX;
const AA: i32 = imgproc::LINE_AA;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Action {
    None,
    Quit,
    Recalibrate,
    ToggleSend,
}

pub struct Overlay {
    window: Window,
    buf: Vec<u32>,
    show_mask: bool,
    last_t: Option<Instant>,
    fps: f64,
}

impl Overlay {
    pub fn new() -> Result<Self, minifb::Error> {
        let w = config::FRAME_WIDTH as usize;
        let h = config::FRAME_HEIGHT as usize;
        let mut window = Window::new(
            "Webcam-Steuerung (Beweis-Screen)",
            w,
            h,
            WindowOptions { resize: true, ..WindowOptions::default() },
        )?;
        window.set_position(config::MONITOR2_X_OFFSET as isize, 0);
        // ~60 FPS Begrenzung der Fenster-Updates
        window.set_target_fps(60);
        Ok(Self {
            window,
            buf: vec![0; w * h],
            show_mask: false,
            last_t: None,
            fps: 0.0,
        })
    }

    pub fn is_open(&self) -> bool {
        self.window.is_open()
    }

    /// Fenster wird beim Drop geschlossen; hier nur zur API-Symmetrie.
    pub fn close(&self) {}

    fn base_image(&self, frame: &Mat, active: &[bool]) -> opencv::Result<Mat> {
        if self.show_mask {
            let rows = config::GRID_ROWS;
            let cols = config::GRID_COLS;
            let mut small =
                Mat::new_rows_cols_with_default(rows, cols, core::CV_8UC1, Scalar::all(0.0))?;
            for row in 0..rows {
                for col in 0..cols {
                    if active[(row * cols + col) as usize] {
                        *small.at_2d_mut::<u8>(row, col)? = 255;
                    }
                }
            }
            let mut big = Mat::default();
            imgproc::resize(&small, &mut big, Size::new(frame.cols(), frame.rows()),
                0.0, 0.0, imgproc::INTER_NEAREST)?;
            let mut bgr = Mat::default();
            imgproc::cvt_color_def(&big, &mut bgr, imgproc::COLOR_GRAY2BGR)?;
            Ok(bgr)
        } else {
            frame.try_clone()
        }
    }

    /// Zeichnet alles auf den Frame und gibt das fertige BGR-Bild zurueck.
    fn render(
        &mut self,
        frame: &Mat,
        active: &[bool],
        calibrated: bool,
        current_key: Option<Key>,
        triggered_zone: Option<usize>,
        send_active: bool,
    ) -> opencv::Result<Mat> {
        let mut img = self.base_image(frame, active)?;
        let w = img.cols();
        let h = img.rows();
        let cols = config::GRID_COLS;
        let rows = config::GRID_ROWS;

        // aktive Zellen halbtransparent
        let mut layer = img.try_clone()?;
        let cw = w as f64 / cols as f64;
        let ch = h as f64 / rows as f64;
        for row in 0..rows {
            for col in 0..cols {
                if active[(row * cols + col) as usize] {
                    let p0 = Point::new((col as f64 * cw) as i32, (row as f64 * ch) as i32);
                    let p1 = Point::new(((col + 1) as f64 * cw) as i32, ((row + 1) as f64 * ch) as i32);
                    imgproc::rectangle_points(&mut layer, p0, p1, Scalar::new(0.0, 200.0, 255.0, 0.0), -1, AA, 0)?;
                }
            }
        }
        let mut blended = Mat::default();
        core::add_weighted(&layer, 0.35, &img, 0.65, 0.0, &mut blended, -1)?;
        img = blended;

        // Rasterlinien
        let grid_col = Scalar::new(60.0, 60.0, 60.0, 0.0);
        for c in 1..cols {
            let x = c * w / cols;
            imgproc::line(&mut img, Point::new(x, 0), Point::new(x, h), grid_col, 1, AA, 0)?;
        }
        for r in 1..rows {
            let y = r * h / rows;
            imgproc::line(&mut img, Point::new(0, y), Point::new(w, y), grid_col, 1, AA, 0)?;
        }

        // Zonenrahmen + getriggerte Zone
        let names = ["left", "right", "up", "down"];
        for (i, rect) in config::ZONE_RECTS.iter().enumerate() {
            let p0 = Point::new((rect.x0 * w as f64) as i32, (rect.y0 * h as f64) as i32);
            let p1 = Point::new((rect.x1 * w as f64) as i32, (rect.y1 * h as f64) as i32);
            let triggered = triggered_zone == Some(i);
            let color = if triggered { Scalar::new(0.0, 0.0, 255.0, 0.0) } else { Scalar::new(0.0, 255.0, 0.0, 0.0) };
            let thick = if triggered { 4 } else { 1 };
            imgproc::rectangle_points(&mut img, p0, p1, color, thick, AA, 0)?;
            imgproc::put_text(&mut img, names[i], Point::new(p0.x + 4, p0.y + 18), FONT, 0.5, color, 1, AA, false)?;
        }

        // grosse Tastenanzeige
        if let Some(k) = current_key {
            let label = match k {
                Key::A => "<- A",
                Key::D => "D ->",
                Key::Space => "[ SPRUNG ]",
                Key::S => "v S",
            };
            imgproc::put_text(&mut img, label, Point::new((w as f64 * 0.30) as i32, 50), FONT, 1.4,
                Scalar::new(0.0, 0.0, 255.0, 0.0), 3, AA, false)?;
        }

        // SEND-Status oben rechts
        let (txt, col) = if send_active {
            ("SEND: AN", Scalar::new(0.0, 0.0, 255.0, 0.0))
        } else {
            ("SEND: AUS", Scalar::new(180.0, 180.0, 180.0, 0.0))
        };
        let mut baseline = 0;
        let ts = imgproc::get_text_size(txt, FONT, 0.9, 2, &mut baseline)?;
        imgproc::put_text(&mut img, txt, Point::new(w - ts.width - 12, 36), FONT, 0.9, col, 2, AA, false)?;

        // FPS
        let now = Instant::now();
        if let Some(prev) = self.last_t {
            let dt = now.duration_since(prev).as_secs_f64();
            if dt > 0.0 {
                self.fps = 0.9 * self.fps + 0.1 * (1.0 / dt);
            }
        }
        self.last_t = Some(now);

        let e = config::ZONE_ENTER_RATIOS;
        let x = config::ZONE_EXIT_RATIOS;
        let status = format!(
            "FPS {:4.1} | Grid {}x{} | enter L{:.2} R{:.2} U{:.2} D{:.2} exit L{:.2} R{:.2} U{:.2} D{:.2} | {} | m=Maske c=Kalib k=Send q=Ende",
            self.fps, cols, rows, e[0], e[1], e[2], e[3], x[0], x[1], x[2], x[3],
            if calibrated { "KALIBRIERT" } else { "kalibriere..." }
        );
        imgproc::put_text(&mut img, &status, Point::new(6, h - 8), FONT, 0.45,
            Scalar::new(255.0, 255.0, 255.0, 0.0), 1, AA, false)?;

        Ok(img)
    }

    /// BGR-Mat -> u32-ARGB-Puffer fuer minifb.
    fn blit(&mut self, img: &Mat) -> opencv::Result<(usize, usize)> {
        let w = img.cols() as usize;
        let h = img.rows() as usize;
        if self.buf.len() != w * h {
            self.buf.resize(w * h, 0);
        }
        let bytes = img.data_bytes()?; // BGR, row-major (Mat ist continuous)
        for i in 0..(w * h) {
            let b = bytes[3 * i] as u32;
            let g = bytes[3 * i + 1] as u32;
            let r = bytes[3 * i + 2] as u32;
            self.buf[i] = (r << 16) | (g << 8) | b;
        }
        Ok((w, h))
    }

    pub fn show(
        &mut self,
        frame: &Mat,
        active: &[bool],
        calibrated: bool,
        current_key: Option<Key>,
        triggered_zone: Option<usize>,
        send_active: bool,
    ) -> opencv::Result<Action> {
        if !frame.empty() {
            let img = self.render(frame, active, calibrated, current_key, triggered_zone, send_active)?;
            let (w, h) = self.blit(&img)?;
            let _ = self.window.update_with_buffer(&self.buf, w, h);
        } else {
            self.window.update();
        }

        if !self.window.is_open() || self.window.is_key_down(MKey::Escape) || self.window.is_key_down(MKey::Q) {
            return Ok(Action::Quit);
        }
        for key in self.window.get_keys_pressed(KeyRepeat::No) {
            match key {
                MKey::M => self.show_mask = !self.show_mask,
                MKey::C => return Ok(Action::Recalibrate),
                MKey::K => return Ok(Action::ToggleSend),
                _ => {}
            }
        }
        Ok(Action::None)
    }
}
